cpp/src/Glacier2Lib/NullPermissionsVerifier.cpp

```diff
@@ -26,114 +26,88 @@ class NullSSLPermissionsVerifier : public Glacier2::SSLPermissionsVerifier
-    for(vector<string>::const_iterator p = props.begin(); p != props.end(); ++p)
+    shared_ptr<Glacier2::PermissionsVerifier> nullPermissionsVerifier;
+    shared_ptr<Glacier2::SSLPermissionsVerifier> nullSSLPermissionsVerifier;
+
+    for (const auto& propertyName : permissionsVerifierPropertyNames)
     {
-        string val = properties->getProperty(*p);
-        if(!val.empty())
+        string val = properties->getProperty(propertyName);
+        if (!val.empty())
         {
-            //
-            // Check permission verifier proxy. It returns a non-empty
-            // value with the new stringified proxy if the property
-            // needs to be rewritten.
-            //
-            val = checkPermissionVerifier(val);
-            if(!val.empty())
+            try
             {
-                properties->setProperty(*p, val);
+                ObjectPrx prx(communicator, val);
+                if (prx->ice_getIdentity() == nullPermissionsVerifierId && nullPermissionsVerifier == nullptr)
+                {
+                    setupObjectAdapter(communicator, adapter);
+                    nullPermissionsVerifier = make_shared<NullPermissionsVerifier>();
+                    adapter->add(nullPermissionsVerifier, nullPermissionsVerifierId);
+                }
+                else if (prx->ice_getIdentity() == nullSSLPermissionsVerifierId && nullSSLPermissionsVerifier == nullptr)
+                {
+                    setupObjectAdapter(communicator, adapter);
+                    nullSSLPermissionsVerifier = make_shared<NullSSLPermissionsVerifier>();
+                    adapter->add(nullSSLPermissionsVerifier, nullSSLPermissionsVerifierId);
+                }
+            }
+            catch(const ProxyParseException&)
+            {
+                // check if it's actually a stringified identity
+                // (with typically missing " " because the category contains a space)
+
+                if(val == communicator->identityToString(nullPermissionsVerifierId) && nullPermissionsVerifier == nullptr)
+                {
+                    setupObjectAdapter(communicator, adapter);
+                    nullPermissionsVerifier = make_shared<NullPermissionsVerifier>();
+                    ObjectPrx prx = adapter->add(nullPermissionsVerifier, nullPermissionsVerifierId);
+                    properties->setProperty(propertyName, prx->ice_toString());
+                }
+                else if(val == communicator->identityToString(nullSSLPermissionsVerifierId) && nullSSLPermissionsVerifier == nullptr)
+                {
+                    setupObjectAdapter(communicator, adapter);
+                    nullSSLPermissionsVerifier = make_shared<NullSSLPermissionsVerifier>();
+                    ObjectPrx prx = adapter->add(nullSSLPermissionsVerifier, nullSSLPermissionsVerifierId);
+                    properties->setProperty(propertyName, prx->ice_toString());
```

Comment:
Shouldn't these still be set even if nullPermisionsVerifier is not null?

Reply:
Yes, you are right!